src/Microsoft.Identity.Web/TokenAcquisition.cs

```diff
@@ -312,36 +338,56 @@ private IConfidentialClientApplication BuildConfidentialClientApplication()
                 request.PathBase,
                 microsoftIdentityOptions.CallbackPath.Value ?? string.Empty);
 
-            if (!applicationOptions.Instance.EndsWith("/"))
-                applicationOptions.Instance += "/";
-
-            string authority ;
-            IConfidentialClientApplication app = null;
-
-            if (microsoftIdentityOptions.IsB2C)
+           if (string.IsNullOrEmpty(applicationOptions.Instance) &&
+                !string.IsNullOrEmpty(microsoftIdentityOptions.Instance))
             {
-                authority = $"{applicationOptions.Instance}tfp/{microsoftIdentityOptions.Domain}/{microsoftIdentityOptions.DefaultUserFlow}";
-                app = ConfidentialClientApplicationBuilder
-                    .CreateWithApplicationOptions(applicationOptions)
-                    .WithRedirectUri(currentUri)
-                    .WithB2CAuthority(authority)
-                    .Build();
+                if (!microsoftIdentityOptions.Instance.EndsWith("/"))
+                    microsoftIdentityOptions.Instance += "/";
+                instance = microsoftIdentityOptions.Instance;
             }
             else
             {
-                authority = $"{applicationOptions.Instance}{applicationOptions.TenantId}/";
-                app = ConfidentialClientApplicationBuilder
-                    .CreateWithApplicationOptions(applicationOptions)
-                    .WithRedirectUri(currentUri)
-                    .WithAuthority(authority)
-                    .Build();
+                if (!applicationOptions.Instance.EndsWith("/"))
+                    applicationOptions.Instance += "/";
+                instance = applicationOptions.Instance;
             }
 
-            // Initialize token cache providers
-            _tokenCacheProvider?.InitializeAsync(app.AppTokenCache);
-            _tokenCacheProvider?.InitializeAsync(app.UserTokenCache);
+            string authority;
+            IConfidentialClientApplication app;
 
-            return app;
+            try
+            {
+                if (microsoftIdentityOptions.IsB2C)
+                {
+                    authority = $"{instance}tfp/{microsoftIdentityOptions.Domain}/{microsoftIdentityOptions.DefaultUserFlow}";
+                    app = ConfidentialClientApplicationBuilder
+                        .CreateWithApplicationOptions(applicationOptions)
+                        .WithRedirectUri(currentUri)
+                        .WithB2CAuthority(authority)
+                        .Build();
+                }
+                else
+                {
+                    authority = $"{instance}{applicationOptions.TenantId}/";
+
+                    app = ConfidentialClientApplicationBuilder
+                        .CreateWithApplicationOptions(applicationOptions)
+                        //.WithRedirectUri(currentUri)
```

Comment:
>//.WithRedirectUri(currentUri) [](start = 24, length = 30)

This is another issue. As we calculate the redirectUri here and then use it, if it's not registered, we fail with a null ref. Issue tracked here: [https://github.com/AzureAD/microsoft-identity-web/issues/38](https://github.com/AzureAD/microsoft-identity-web/issues/38)

Reply:
working around this for now by creating the httpcontext, but needs to be looked into (hence the issue)

---
In reply to: [385444973](https://github.com/AzureAD/microsoft-identity-web/pull/39#discussion_r385444973) [](ancestors = 385444973)